### ISSUE TITLE
Sample says COMPLETE but did not get indexed on portal. 

### ISSUE THREAD

**Issue by rosibaj** [cancogen-virus-seq/portal#95]
# Description of Issue 
User has reported a sample is listed as COMPLETE on upload, but did not make it to the portal: 
![image](https://user-images.githubusercontent.com/15824252/133340401-6564574e-b142-4b2d-89fe-14bfa7d88daa.png)
 

User re-uploaded the sample a second time, and this one is now visible: 

![image](https://user-images.githubusercontent.com/15824252/133340446-7382b2a2-b57c-43f4-98fa-3959f5bd5711.png)

# Analysis of Data

## Muse upload 
submission id - `a03404b8-1358-4cae-8073-87f209e56f17`
https://muse.virusseq-dataportal.ca/submissions/a03404b8-1358-4cae-8073-87f209e56f17 

## Song analysis
The analysis that is listed as COMPLETE is PUBLISHED in Song. 
```
curl --location --request GET 'https://song.virusseq-dataportal.ca/studies/BCCDC-BC/analysis/ae228aa2-f2ca-4333-a28a-a2f2ca13333a' \
```
![image](https://user-images.githubusercontent.com/15824252/133339319-d3398993-8d6b-46a4-a4ac-cb6ceab1eb3c.png)

## Query to arranger 
When i search for this analysis against the lastest Arranger metadata API, i get no results.  I can find the second uplaod: 

### No result query  

```

POST /cancogen_sep10/graphql HTTP/1.1
Host: arranger.virusseq-dataportal.ca
Content-Type: application/json
Content-Length: 645

{"query":"query ($sqon: JSON) {\r\n  file {\r\n    hits(filters: $sqon) {\r\n      edges {\r\n        node {\r\n          analysis{\r\n            analysis_id\r\n            first_published_at\r\n          }\r\n          donors {\r\n            hits {\r\n              edges {\r\n                node {\r\n                  submitter_donor_id\r\n                  donor_id\r\n                }\r\n              }\r\n            }\r\n          }\r\n        }\r\n      }\r\n    }\r\n  }\r\n}\r\n","variables":{"sqon":{"op":"and","content":[{"op":"in","content":{"field":"analysis.analysis_id","value":["ae228aa2-f2ca-4333-a28a-a2f2ca13333a"]}}]}}}
```
![image](https://user-images.githubusercontent.com/15824252/133339213-8523e31b-365a-4ab9-9157-3a6eea66b0d2.png)

### 2nd upload query 
```
POST /cancogen_sep10/graphql HTTP/1.1
Host: arranger.virusseq-dataportal.ca
Content-Type: application/json
Content-Length: 645

{"query":"query ($sqon: JSON) {\r\n  file {\r\n    hits(filters: $sqon) {\r\n      edges {\r\n        node {\r\n          analysis{\r\n            analysis_id\r\n            first_published_at\r\n          }\r\n          donors {\r\n            hits {\r\n              edges {\r\n                node {\r\n                  submitter_donor_id\r\n                  donor_id\r\n                }\r\n              }\r\n            }\r\n          }\r\n        }\r\n      }\r\n    }\r\n  }\r\n}\r\n","variables":{"sqon":{"op":"and","content":[{"op":"in","content":{"field":"analysis.analysis_id","value":["f3cdc94d-d92e-4d58-8dc9-4dd92e6d5859"]}}]}}}
```
![image](https://user-images.githubusercontent.com/15824252/133340691-2bd8345f-d54c-471f-98a4-f297f4698481.png)

**Comment by rosibaj** [comment]
- analysis that is indexed `f3cdc94d-d92e-4d58-8dc9-4dd92e6d5859` 
- analysis that is not indexed `ae228aa2-f2ca-4333-a28a-a2f2ca13333a'

**Comment by lepsalex** [comment]
![Screenshot from 2021-09-16 15-44-43](https://user-images.githubusercontent.com/3143103/133677645-098d2d60-89a5-487f-90d1-7c4e6c8ca8ba.png)

Ok so this was a Song error, as you can see from the screenshot:

* the first two logs entries show that the analysis was published, event emitted by kafka, received by maestro
* the rest of entries show that Maestro was attempting to get this analysis from Song in order to publish it to ES and that request to SONG resulted in a 500 (internal server error), it retried a couple times and then eventually gave up
* Song logs show the request coming in 3 times (initial plus retries) 
* Once maestro exhausted it's retries it gave up and moved on

That analysis is there now as you can see by going [here](https://song.virusseq-dataportal.ca/studies/BCCDC-BC/analysis/ae228aa2-f2ca-4333-a28a-a2f2ca13333a)

Looking at Song logs for the same time as these 500 errors ...

![Screenshot from 2021-09-16 16-00-18](https://user-images.githubusercontent.com/3143103/133678018-e63a2278-3f45-44d3-9d74-a11a6f9453c1.png)

We can see that this instance of Song was unable to get a connection to the database. It is strange that only this one GET request failed while the rest of the submission was fine.

#### TLDR Conclusion

Song database connections got overloaded and resulted in a 500 error from an instance of song as it attempted to fetch the analysis for maestro to index.

#### Recommendations

We can look at potentially tweaking some of the configs to allow for more connections and/or a beefier postgres but we may also want to observer if this happens again what the size of the submission is to understand if this is something that scales with submission size.

**Comment by rosibaj** [comment]
@lepsalex looks good - following up with @yalturmes now about possible alerts in zabbix.  will close once i know outcome of that.

**Comment by rosibaj** [comment]
@lindaxiang can you suppress the duplicate sample of analysis `a03404b8-1358-4cae-8073-87f209e56f17` on production so wwe can close this ticket out?

**Comment by rosibaj** [comment]
Three more observed instances: 

curl --location --request GET 'https://song.virusseq-dataportal.ca/studies/BCCDC-BC/analysis/08264d05-e0e8-43ab-a64d-05e0e813abd9'

- BC_169005 ed0268a9-90b8-426a-8268-a990b8d26ac5 COMPLETE
- BC_169017 e2c899fe-7e82-4555-8899-fe7e8245557f COMPLETE
- BC_172062 08264d05-e0e8-43ab-a64d-05e0e813abd9 COMPLETE
  - PUBLISHED analysis: https://song.virusseq-dataportal.ca/studies/BCCDC-BC/analysis/08264d05-e0e8-43ab-a64d-05e0e813abd9'

**Comment by rosibaj** [comment]
Moving back to in-progress to discuss how to resovle in the next sprint planning

**Comment by rosibaj** [comment]
Discussion in sprint planning: 

Solution has been spec'd: https://wiki.oicr.on.ca/display/OV/Paging+in+Song#PaginginSong-SQLOptimisation 

Fixing this issue depends on a couple options: 
1. Update to optimize these queries to/from song.  Volume in VirusSeq is 
2. Maestro should index from the Analysis https://github.com/overture-stack/maestro/issues/167

**Comment by nithujohn** [comment]
@b-f-chan @lindaxiang  Samples are not indexing on the portal. Attached the file with analysis ID and specimen ID 
[PHO IDs not indexed.xlsx](https://github.com/cancogen-virus-seq/portal/files/7392612/PHO.IDs.not.indexed.xlsx)

**Comment by nithujohn** [comment]
> Three more observed instances:
> 
> * BC_169005 ed0268a9-90b8-426a-8268-a990b8d26ac5 COMPLETE
> * BC_169017 e2c899fe-7e82-4555-8899-fe7e8245557f COMPLETE
> * BC_172062 08264d05-e0e8-43ab-a64d-05e0e813abd9 COMPLETE
> 
> All are PUBLISHED analysis - e.g. https://song.virusseq-dataportal.ca/studies/BCCDC-BC/analysis/08264d05-e0e8-43ab-a64d-05e0e813abd9'
> 
> However, they were not indexed:
> 
> ![image](https://user-images.githubusercontent.com/15824252/134962235-fc20427e-3825-463a-83ad-58a85be3119c.png)
> 
> ```
> query ($sqon: JSON) {
>   file {
>     hits(filters: $sqon) {
>       edges {
>         node {
>           analysis {
>             analysis_id
>             first_published_at
>           }
>           donors {
>             hits {
>               edges {
>                 node {
>                   submitter_donor_id
>                   donor_id
>                 }
>               }
>             }
>           }
>         }
>       }
>     }
>   }
> }
> ```
> 
> ```
> {
> 	"sqon": {
> 		"op": "and",
> 		"content": [{
> 			"op": "in",
> 			"content": {
> 				"field": "analysis.analysis_id",
> 				"value": ["ae228aa2-f2ca-4333-a28a-a2f2ca13333a","ed0268a9-90b8-426a-8268-a990b8d26ac5","e2c899fe-7e82-4555-8899-fe7e8245557f","08264d05-e0e8-43ab-a64d-05e0e813abd9"]
> 			}
> 		}]
> 	}
> }
> ```

@scottcain @b-f-chan Can we get these three samples indexed as well?

**Comment by jaserud** [comment]
@scottcain  As discussed I've split this ticket into the separate issues that we found:
- https://github.com/cancogen-virus-seq/portal/issues/147
- https://github.com/cancogen-virus-seq/roadmap/issues/32

All issues regarding samples not indexed in this ticket should already be resolved. If there are any other indexing issues, let's open new tickets.
